### PR TITLE
Add CTC objective function in optax.loss.

### PR DESCRIPTION
**Pull request by yotarok**
CTC objective function is added to loss.py.

### CODE REVIEW

File: optax/_src/loss.py [hunk 1]
```diff
@@ -252,3 +252,99 @@ def log_cosh(
...
+    labels: (B, N)-array containing reference integer labels.
+    labelpaddings: (B, N)-array. Padding indicators for `labels`. Currently,
+      `labels` must be right-padded, i.e. each row of `labelpaddings` must be
+      repetition of zeroes, followed by repetition of ones.
+    blank_id: Id for blank token.
+
+  Returns:
+    (B,)-array containing loss values for each sequence in the batch.
+  """
+  batchsize, unused_maxinputlen, num_classes = logits.shape
```
Comment (hbq1) [inline optax/_src/loss.py:286]: Consider calling `chex.assert_rank(logits, 3)` for more informative exceptions

File: optax/_src/loss.py [hunk 2]
```diff
@@ -252,3 +252,99 @@ def log_cosh(
...
+  batchsize_of_labels, maxlabellen = labels.shape
```
Comment (hbq1) [inline optax/_src/loss.py:287]: Same here

File: optax/_src/loss.py [hunk 3]
```diff
@@ -252,3 +252,99 @@ def log_cosh(
...
+      (batchsize, maxlabellen + 1)) * logepsilon  # [B, N]
+  logalpha_phi_init = logalpha_phi_init.at[:, 0].set(0.0)
+  logalpha_emit_init = jnp.ones(
+      (batchsize, maxlabellen)) * logepsilon  # [B, N]
+
+  def loop_body(prev, x):
+    prev_phi, prev_emit = prev
+    # emit-to-phi epsilon transition, except if the next label is repetition
+    prev_phi_orig = prev_phi
+    prev_phi = prev_phi.at[:, 1:].set(
```
Comment (hbq1) [inline optax/_src/loss.py:314]: Out of curiosity, did you run any benchmarks to measure the efficiency of this operation? I heard that .set() is very slow, and, since you're updating a contiguous subtensor, a simple concatenation might work a way faster.

E.g. you could have a helper function

```
def _update_1up(phi, update):
   return jnp.concatenate([phi[:, :1], update], axis=-1)
```

File: optax/_src/loss.py [hunk 4]
```diff
@@ -252,3 +252,99 @@ def log_cosh(
...
+
+    logprob_emit, logprob_phi, pad = x
+
+    # phi-to-emit transition
+    next_emit = jnp.logaddexp(prev_phi[:, :-1] + logprob_emit,
+                              prev_emit + logprob_emit)
+    # self-loop transition
+    next_phi = prev_phi + logprob_phi
+    # emit-to-phi blank transition only when the next label is repetition
+    next_phi = next_phi.at[:, 1:].set(
```
Comment (hbq1) [inline optax/_src/loss.py:325]: Same here

File: optax/_src/loss.py (hunk 1 above)
Comment (yotarok) [reply optax/_src/loss.py:295]: Thanks, done.

File: optax/_src/loss.py (hunk 2 above)
Comment (yotarok) [reply optax/_src/loss.py:296]: done.

File: optax/_src/loss.py (hunk 3 above)
Comment (yotarok) [reply optax/_src/loss.py]: Thanks for the suggestion, done.
In my benchmark script, it improved the performance by >30%,
specifically from ~27ms to ~20ms in the setting with `logits.shape==(128, 1024, 64)`, and `labels.shape==(128, 256)`.

File: optax/_src/loss.py (hunk 4 above)
Comment (yotarok) [reply optax/_src/loss.py]: done

File: optax/_src/loss.py [hunk 5]
```diff
@@ -252,3 +252,112 @@ def log_cosh(
...
+
+    return (next_phi, next_emit), (next_phi, next_emit)
+
+  xs = (logprobs_emit, logprobs_phi, logitpaddings.transpose((1, 0)))
+  _, (logalpha_phi,
+      logalpha_emit) = jax.lax.scan(loop_body,
+                                    (logalpha_phi_init, logalpha_emit_init), xs)
+
+  # last row needs to be updated with the last epsilon transition
+  logalpha_phi_last = logalpha_phi[-1].at[:, 1:].set(
```
Comment (hbq1) [inline optax/_src/loss.py:354]: `update_phi_score(logalpha_phi[-1], ...))`?

Comment (yotarok) [reply optax/_src/loss.py]: Great catch. Thanks, done.
And, I noticed that the next line is a left over from my previous version that returns "logalpha"s as well as "per_seq_loss".
Since this function doesn't need to compute accurate "logalpha_phi", line 355 was deleted.

**Review by hbq1 — APPROVED** [review]

File: optax/_src/loss_test.py [hunk 6]
```diff
@@ -238,5 +239,107 @@ def test_batched_predictions_only(self):
...
+      for t in range(steps):
+        p += logprobs[b, t, labels[b, t]]
+      np.testing.assert_allclose(jnp.array(-p), per_seq_loss[b],
+                                 rtol=1e-6)
+
+  @chex.all_variants
+  def test_with_one_to_one_alignment_and_paddings(self):
+    # test if padding indicators work as expected.
+    batch_size = 1
+    nclasses = 8
```
Comment (wchan) [inline optax/_src/loss_test.py:294]: for additional test coverage, can we expand this to batch_size == 7 or an odd number), nclasses == 33 (or an odd number), steps == 5

File: optax/_src/loss_test.py [hunk 7]
```diff
@@ -238,5 +239,107 @@ def test_batched_predictions_only(self):
...
+    label_padding = np.array([[0, 0, 0, 1]], dtype=np.float32)
+
+    value_and_grad = self.variant(jax.value_and_grad(_average_ctc_loss))
+    actual_loss, actual_gradients = value_and_grad(
+        logits, logit_padding, labels, label_padding)
+    expected_loss = -(logprobs[0, 0, 1] + logprobs[0, 1, 2] + logprobs[0, 2, 3])
+    np.testing.assert_allclose(expected_loss, actual_loss)
+
+    expected_gradients = np.array(jax.nn.softmax(logits))
+    expected_gradients[0, 0, 1] -= 1.0
```
Comment (wchan) [inline optax/_src/loss_test.py:309]: can we compute the gradients with jax (instead of manually computing them)

File: optax/_src/loss_test.py [hunk 8]
```diff
@@ -238,5 +239,107 @@ def test_batched_predictions_only(self):
...
+
+  @chex.all_variants
+  def test_with_one_to_one_alignment_and_paddings(self):
+    # test if padding indicators work as expected.
+    batch_size = 1
+    nclasses = 8
+    steps = 4
+    logits = np.random.normal(size=[batch_size, steps, nclasses])
+    logprobs = jax.nn.log_softmax(logits)
+    labels = np.array([[1, 2, 3, 4]])
```
Comment (wchan) [inline optax/_src/loss_test.py:298]: maybe make the labels randomly constructed like:

// something like this, sample w/o repetitions
range(1, nclasses)
shuffle()
truncate [:steps]

File: optax/_src/loss_test.py [hunk 9]
```diff
@@ -238,5 +239,107 @@ def test_batched_predictions_only(self):
...
+    expected_gradients[0, 2, 3] -= 1.0
+    expected_gradients[0, 3, :] = 0.0
+    np.testing.assert_allclose(expected_gradients, actual_gradients,
+                               rtol=1e-6)
+
+  @chex.all_variants
+  def test_repeat_with_one_to_one_alignment(self):
+    # test if it can correctly handle the same-label repetition.
+    batch_size = 2
+    labels = np.array([
```
Comment (wchan) [inline optax/_src/loss_test.py:320]: can we add a few more cases, where the labels is like

[1, 1, 1, 1]
[1, 1, 2, 3]
[1, 1, 1,, 2]

File: optax/_src/loss_test.py [hunk 10]
```diff
@@ -238,5 +239,107 @@ def test_batched_predictions_only(self):
...
+        [2, 3, 4, 0, 4],
+    ]
+
+    for n in range(batch_size):
+      expected_loss = -sum(logprobs[n, t, k]
+                           for t, k in enumerate(expected_alignment[n]))
+      np.testing.assert_allclose(
+          jnp.array(expected_loss), per_seq_loss[n], rtol=1e-6)
+
 if __name__ == '__main__':
```
Comment (wchan) [inline optax/_src/loss_test.py:344]: optional: consider add a test comparing your implementation w/ TF on both the forward pass and backward pass w various values. the tolerance of the test won't be very good though, so it might not be super helpful.

**Review by wchan — COMMENTED** [review]
this is awesome, thanks @yotarok !

i just have some minor testing coverage comments.

File: optax/_src/loss_test.py (hunk 6 above)
Comment (yotarok) [reply optax/_src/loss_test.py]: done.

File: optax/_src/loss_test.py (hunk 8 above)
Comment (yotarok) [reply optax/_src/loss_test.py]: done.

File: optax/_src/loss_test.py (hunk 9 above)
Comment (yotarok) [reply optax/_src/loss_test.py:332]: done

File: optax/_src/loss_test.py (hunk 7 above)
Comment (yotarok) [reply optax/_src/loss_test.py]: The current gradient computation is already depending on Jax's grad (no custom backward implemented).
So, I think it's more important to compare it with manually-computed values.

File: optax/_src/loss_test.py (hunk 10 above)
Comment (yotarok) [reply optax/_src/loss_test.py:370]: I was just reluctant to add another test dependency package.  Actually, I have such tests already, and if it's okay to add test dependency I can move the code to this repo.

**Comment by yotarok** [comment]
Thanks for the comments. Most of the comments are reflected into the code, so resolved. For the rest of them, I keep it open so please let me know if you can suggest something I can do.
(Especially, I'm curious if I should add tensorflow as a test dependency, and do parity tests against TF.)

File: optax/_src/loss_test.py (hunk 10 above)
Comment (hbq1) [reply optax/_src/loss_test.py:370]: It is a good suggestion! 
We should consider adding parity tests against TF implementations; I'll discuss it with other Optax core devs. In any case, this is out of scope of this PR.

**Review by hbq1 — APPROVED** [review]

**Comment by yotarok** [comment]
Looks like this is already merged by Copybara bot, so I will close this PR.
